### PR TITLE
Honor column dtype from input dataframe when roundtripping

### PR DESCRIPTION
**Pull request by phofl**

### CODE REVIEW

**Comment by phofl** [comment]
No worries, it was my own fault, didn't install pytest-asyncio which skipped the pandas tests that are failing now... I wanna see if this fixes it before we merge anything else.

**Comment by phofl** [comment]
So summary what's going wrong:

pandas changed the default columns dtype when a DataFrame is constructed from an empty dict, this will default to a RangeIndex now. Fastparquet isn't honouring the dtype that is written to the parquet file (part of the problem might have been that it was always object). This PR

- writes the proper columns dtype to the metadata of the parquet file
- reads it back and constructs the column index accordingly

This should solve the problems of not getting back a proper dtype.

The ``"no_columns.parquet"`` seems to be from a time when no columns metadata were provided at all, meaning the test will fail on newer pandas version (that's why it is special cased). Way forward would be: Merge the other pr that gets rid of the simple categorical deprecation warnings and then rebase here to check that the changes make sense, should make reviewing easier as well.

File: fastparquet/api.py [hunk 1]
```diff
@@ -887,6 +888,10 @@ def categories(self):
             return self._categories
         if self.has_pandas_metadata:
             metadata = self.pandas_metadata
+            if "column_indexes" in metadata and len(metadata["column_indexes"]) > 0:
+                self._columns_dtype = metadata["column_indexes"][0]["numpy_type"]
```
Comment (martindurant) [inline fastparquet/api.py:892]: To be sure memory fails): does fastparquet write this metadata? Does arrow?

Comment (phofl) [reply fastparquet/api.py:892]: Yes both do

**Comment by martindurant** [comment]
OK, all good here

**Pull request merged**